dandiapi/api/services/metadata/__init__.py

```diff
@@ -0,0 +1,106 @@
+import jsonschema.exceptions
+
+from dandiapi.api.models import Asset, Version
+
+logger = get_task_logger(__name__)
+
+
+def _encode_pydantic_error(error) -> dict[str, str]:
+    return {'field': error['loc'][0], 'message': error['msg']}
+
+
+def _encode_jsonschema_error(error: jsonschema.exceptions.ValidationError) -> dict[str, str]:
+    return {'field': '.'.join([str(p) for p in error.path]), 'message': error.message}
+
+
+def _collect_validation_errors(
+    error: dandischema.exceptions.ValidationError,
+) -> list[dict[str, str]]:
+    if type(error) is dandischema.exceptions.PydanticValidationError:
+        encoder = _encode_pydantic_error
+    elif type(error) is dandischema.exceptions.JsonschemaValidationError:
+        encoder = _encode_jsonschema_error
+    else:
+        raise error
+    return [encoder(error) for error in error.errors]
+
+
+def validate_asset_metadata(*, asset_id: int) -> None:
+    logger.info('Validating asset metadata for asset %s', asset_id)
+    asset: Asset = Asset.objects.get(id=asset_id)
+
+    # Published assets are immutable
+    assert not asset.published
+
+    with transaction.atomic():
+        asset.status = Asset.Status.VALIDATING
+        asset.save()
+
+        try:
+            metadata = asset.published_metadata()
+            validate(metadata, schema_key='PublishedAsset', json_validation=True)
+            logger.info('Successfully validated asset %s', asset_id)
+            asset.status = Asset.Status.VALID
+            asset.validation_errors = []
+        except dandischema.exceptions.ValidationError as e:
+            logger.info('Error while validating asset %s', asset_id)
+            asset.status = Asset.Status.INVALID
+
+            validation_errors = _collect_validation_errors(e)
+            asset.validation_errors = validation_errors
+        except ValueError as e:
+            # A bare ValueError is thrown when dandischema generates its own exceptions, like a
+            # mismatched schemaVersion.
+            asset.status = Asset.Status.INVALID
+            asset.validation_errors = [{'field': '', 'message': str(e)}]
+
+        # Save asset
+        asset.save()
+
+        # Update modified timestamps on all draft versions this asset belongs to
+        asset.versions.filter(version='draft').update(modified=timezone.now())
+
+
+def validate_version_metadata(*, version_id: int) -> None:
+    logger.info('Validating dandiset metadata for version %s', version_id)
+    version: Version = Version.objects.get(id=version_id)
+
+    # Published versions are immutable
+    assert version.version == 'draft'
+
+    with transaction.atomic():
+        version.status = Version.Status.VALIDATING
+        version.save()
+
+        try:
+            publish_version = version.publish_version
+            metadata = publish_version.metadata
+
+            # Inject a dummy DOI so the metadata is valid
+            metadata['doi'] = '10.80507/dandi.123456/0.123456.1234'
```

Comment:
Is there instead a way to signal to the `validate()` function to ignore a missing DOI?

Reply:
Not in its current form.

I didn't attempt to change any internal behavior of the validation functions for this PR, but eventually I think it's worth revisiting how we're validating version metadata. Right now we basically "fake publish" versions and validate against what the published metadata would be; maybe we should have a way to explicitly validate draft versions.

Reply:
Ok, something to think about later. Moving things to the service layer makes this kind of consideration easier to analyze in the future, so right now I do agree with your minimal-change approach. Thanks!